src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp

```diff
@@ -26,74 +26,232 @@
 #include <lib/support/CHIPMem.h>
 #include <platform/KeyValueStoreManager.h>
 #include <platform/nxp/k32w/k32w0/K32W0Config.h>
+#include <set>
 #include <string>
 
 #include "PDM.h"
 
+#include <unordered_map>
+
 namespace chip {
 namespace DeviceLayer {
 namespace PersistedStorage {
 
-/* TODO: adjust this value */
-#define MAX_NO_OF_KEYS 255
+/* TODO: adjust these values */
+constexpr size_t kMaxNumberOfKeys  = 20;
+constexpr size_t kMaxKeyValueBytes = 255;
 
 KeyValueStoreManagerImpl KeyValueStoreManagerImpl::sInstance;
 
+/* hashmap having:
+ * 	- the matter key string as key;
+ * 	- internal PDM identifier as value;
+ */
+std::unordered_map<std::string, uint8_t> g_kvs_map;
+
+/* set containing used PDM identifiers */
+std::set<uint8_t> g_key_ids_set;
+
+/* used to check if we need to restore values from flash (e.g.: reset) */
+static bool g_restored_from_flash = false;
+
+CHIP_ERROR RestoreFromFlash()
+{
+    CHIP_ERROR err                        = CHIP_NO_ERROR;
+    uint8_t key_id                        = 0;
+    char key_string_id[kMaxKeyValueBytes] = { 0 };
+    size_t key_string_id_size             = 0;
+    uint8_t pdm_id_kvs                    = chip::DeviceLayer::Internal::K32WConfig::kPDMId_KVS;
+    uint16_t pdm_internal_id              = 0;
+
+    if (g_restored_from_flash)
+    {
+        /* already restored from flash, nothing to do */
+        return err;
+    }
+
+    for (key_id = 0; key_id < kMaxNumberOfKeys; key_id++)
+    {
+        /* key was saved as string in flash (key_string_id) using (key_id + kMaxNumberOfKeys) as PDM key
+         * value corresponding to key_string_id was saved in flash using key_id as PDM key
+         */
+
+        pdm_internal_id = chip::DeviceLayer::Internal::K32WConfigKey(pdm_id_kvs, key_id + kMaxNumberOfKeys);
+        err = chip::DeviceLayer::Internal::K32WConfig::ReadConfigValueStr(pdm_internal_id, key_string_id, kMaxKeyValueBytes,
+                                                                          key_string_id_size);
+
+        if (err == CHIP_DEVICE_ERROR_CONFIG_NOT_FOUND)
+        {
+            err = CHIP_NO_ERROR;
+            continue;
+        }
+        else if (err != CHIP_NO_ERROR)
+        {
+            ChipLogProgress(DeviceLayer, "KVS, Error while restoring Matter key [%s] from flash with PDM id: %i", key_string_id,
+                            pdm_internal_id);
+            break;
```

Comment:
```suggestion
            return err;
```

Reply:
Else we set g_restored_from_flash=true, even when init failed.

Reply:
Right, solved.